index.ts

```diff
@@ -0,0 +1,106 @@
+import { FastifyPluginAsync } from "fastify";
+import { sql } from "slonik";
+import { Item, Member, PreHookHandlerType } from "graasp";
+
+const DEFAULT_MAX_STORAGE = 1024 * 1024 * 100; // 100MB;
+
+interface GraaspFileUploadLimiterOptions {
+  /** Item type to target (ex: 'file', 's3File') */
+  type: string;
+  /** Chain of property names to the file size (ex: 'extra.s3File.size')  */
+  sizePath: string;
+  /** Maximum storage size for a user in bytes (default to 100MB)  */
+  maxMemberStorage?: number;
+}
+
+type MemberExtra = {
+  storage: {
+    total: number;
+  };
+};
+
+class StorageExceeded extends Error {
+  message = "The allowed storage is full";
+  name = "StorageExceededError";
+  origin = "plugin";
+  code = 507;
+}
+
+const plugin: FastifyPluginAsync<GraaspFileUploadLimiterOptions> = async (fastify, options) => {
+  const {
+    items: { taskManager },
+    taskRunner: runner,
+    db,
+  } = fastify;
+  const { type: itemType, sizePath, maxMemberStorage = DEFAULT_MAX_STORAGE } = options;
+
+  if (!itemType || !sizePath) {
+    throw new Error("graasp-file-upload-limiter: missing plugin options");
+  }
+
+  const getFileSize = (item): number => {
+    const properties = sizePath.split(".");
+    const sizeField = properties.reduce((prev, curr) => prev?.[curr], item);
+
+    if (Number.isInteger(sizeField)) {
+      return sizeField;
+    }
+
+    throw new Error("sizePath does not lead to a size number");
+  };
+
+  const getMemberStorage = (id: string) => {
+    // todo: fetch total storage from Redis
+
+    // -- the following solution is failing because it cannot extract property from jsonb
+    // const column = sql.join(["extra", "file", "size"], sql`->`);
+    // return db.pool
+    //   .oneFirst<string>(
+    //     sql`
+    //        SELECT
+    //        SUM((${column})) as size
+    //        FROM item
+    //        WHERE item.type ='file' AND item.creator = ${id}`
+    //   )
+    //   .then((size) => parseInt(size, 10) || 0);
+
+    // select all file extra for given member
+    return (
+      db.pool
+        .query<any>(
```

Comment:
I couldn't use `Partial<Item>` or `Item` or `UnknownExtra` here, all of them prevents me from compiling...

Reply:
I resolved it.